### ISSUE TITLE
typevars considered inferable when they shouldn't be

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#1898]
```py
class Container[T]:
    def method(self, x: T) -> T:
        return x

    def try_assign[U](self, x: U) -> U:
        result = self.method(x)  # false negative here
        reveal_type(result)  # revealed: `T@Container`
        return result  # we correctly emit a diagnostic here
```

https://play.ty.dev/8b3bb7c8-d7a7-447a-9318-7ff50e896f90

I think the `self.method(x)` call should be a type error, but currently is not. `U` and `T` are independent type variables, and neither of them is inferable in the context of this call. `T` is already bound because we are in a method of `Container`; `U` is bound because we are inside `try_assign` method.

**Comment by dcreager** [comment]
Possibly solved by https://github.com/astral-sh/ruff/pull/24949

**Comment by charliemarsh-oai** [comment]
Appears to still reproduce.

**Comment by charliermarsh** [comment]
No longer reproduces!
